### PR TITLE
Improve EffectiveTldFinder, fixes #172

### LINKED ISSUE #172: Incorrect domains returned by EffectiveTldFinder
The method `EffectiveTldFinder.getAssignedDomain(hostname)` returns an incorrect domain in a couple of cases. This was detected by a comparison with the Python module [tldextract](https://pypi.python.org/pypi/tldextract) which is also based on the [public suffix list](https://publicsuffix.org/).

1. exceptions to wildcard rules are not recognized as valid domain, e.g. with the rules
   ```
   *.kawasaki.jp
   !city.kawasaki.jp
   ```
   the domain for "www.city.kawasaki.jp" should be "city.kawasaki.jp".

2. the domain detection fails entirely of the .za top-level domain: "blogs.uct.ac.za" should give "ac.za". A test whether the last element "za" is a valid TLD fails but the [.za registrar](https://www.zadna.org.za/content/page/domain-information/) does not allow .za alone as TLD.

3. detection fails for internationalized domain names in their punycoded (ASCII) form: for "спб.бесплатныеобъявления.рф" the domain "бесплатныеобъявления.рф" is correctly returned but not for the punycode equivalent "xn--90a1af.xn--80abbembcyvesfij3at4loa4ff.xn--p1ai". The public suffix list contains the utf-8 representation of IDNs and must be (additionally) transposed to the ASCII/Punycode form.

### PR DESCRIPTION
**Pull request by sebastian-nagel**
- addressed all issues in #172, except IDNs

### CODE REVIEW

File: src/main/java/crawlercommons/domains/EffectiveTldFinder.java [hunk 1]
```diff
@@ -65,26 +96,73 @@
...
+    /**
+     * (Re)initialize EffectiveTldFinder with built-in public suffix list.
+     * 
+     * @param excludePrivateDomains
+     *            whether to exclude the public suffixes listed in the PRIVATE
+     *            domain section (opposed to &quot;ICANN&quot; domains)
+     * @return true if (re)initialization was successful
+     */
+    public boolean initialize(boolean excludePrivateDomains) {
+        return initialize(this.getClass().getResourceAsStream(ETLD_DATA), false);
```
Comment (kkrugler) [inline src/main/java/crawlercommons/domains/EffectiveTldFinder.java]: Shouldn't this pass `excludePrivateDomains` to the `initialize()` call, versus always passing `false`?

File: src/main/java/crawlercommons/domains/EffectiveTldFinder.java [hunk 2]
```diff
@@ -65,26 +96,73 @@
...
+        boolean inPrivateDomainSection = false;
         try {
-            if (null == effectiveTldDataStream && null != this.getClass().getResource(ETLD_DATA)) {
-                effectiveTldDataStream = this.getClass().getResourceAsStream(ETLD_DATA);
-            }
             BufferedReader input = new BufferedReader(new InputStreamReader(effectiveTldDataStream, StandardCharsets.UTF_8));
             String line = null;
             while (null != (line = input.readLine())) {
-                if (line.length() == 0 || (line.length() > 1 && line.startsWith(COMMENT))) {
+                if (line.length() == 0) {
```
Comment (kkrugler) [inline src/main/java/crawlercommons/domains/EffectiveTldFinder.java]: Typically you want to trim the line before checking for empty lines or whether they start with a comment. Also `line.isEmpty()` is better here.

File: src/test/java/crawlercommons/domains/EffectiveTldFinderTest.java [hunk 3]
```diff
@@ -136,11 +140,53 @@ public final void testCoJpDomain() throws Exception {
...
+    @Test
+    public final void testIDNDomain() throws Exception {
+        String ad = null;
+        ad = EffectiveTldFinder.getAssignedDomain("спб.бесплатныеобъявления.рф");
+        // assertEquals("бесплатныеобъявления.рф", ad); // TODO
+        ad = EffectiveTldFinder.getAssignedDomain("xn--90a1af.xn--80abbembcyvesfij3at4loa4ff.xn--p1ai");
+        assertEquals("xn--80abbembcyvesfij3at4loa4ff.xn--p1ai", ad);
+        // rare but possible mixed use of UTF-8 and Punycode
+        ad = EffectiveTldFinder.getAssignedDomain("xn--90a1af.бесплатныеобъявления.рф");
+        // assertEquals("xn--80abbembcyvesfij3at4loa4ff.xn--p1ai", ad); // TODO
```
Comment (kkrugler) [inline src/test/java/crawlercommons/domains/EffectiveTldFinderTest.java]: Add link to new issue for commented out lines?

File: src/main/java/crawlercommons/domains/EffectiveTldFinder.java (hunk 1 above)
Comment (sebastian-nagel) [reply src/main/java/crawlercommons/domains/EffectiveTldFinder.java]: Sure, I'll update the PR.

File: src/main/java/crawlercommons/domains/EffectiveTldFinder.java (hunk 2 above)
Comment (sebastian-nagel) [reply src/main/java/crawlercommons/domains/EffectiveTldFinder.java]: agreed

File: src/test/java/crawlercommons/domains/EffectiveTldFinderTest.java (hunk 3 above)
Comment (sebastian-nagel) [reply src/test/java/crawlercommons/domains/EffectiveTldFinderTest.java]: ok

**Comment by sebastian-nagel** [comment]
PR updated, addressing @kkrugler comments. Thanks!

**Comment by sebastian-nagel** [comment]
Thanks for the review!

**Pull request merged**